static/js/containers/SearchPage.js

```diff
@@ -171,6 +184,11 @@ export class SearchPage extends React.Component<Props, State> {
       )
     }
 
+    if (upvotedPost && upvotedPosts.get(upvotedPost.id) !== upvotedPost) {
+      upvotedPosts.set(upvotedPost.id, upvotedPost)
+      this.setState({ upvotedPosts })
+    }
```

Comment:
I think with this approach it is possible to run into a race condition if the user upvotes / downvotes a whole bunch of posts very quickly on a slow connection. I'm not exactly sure what is causing it, but if I use chrome's connection throttler to slow down the connection a ton I can click upvote on like five or six posts a whole bunch and the UI sometimes (not always) gets out of sync with the actual upvoted / not upvoted state.

I think the source of the issue is that the `data` property on the post upvote reducer (which is used to get the `upvotedPost` value here) only seems to store the value of the most recently clicked one, I think when there are a bunch of requests going off simultaneously it's possible for this value to change in between re-renders of the react component, so one of the posts that should be added to the `upvotedPosts` map is dropped on the floor. At least that's what I think might be going on, I'm not sure.

I'm not sure the best way to get around this, or if we think it will be a huge issue. One solution might be to add an instance method here that fires off the request with the `toggleUpvote` function and then adds the post to the `upvotePosts` map.

One other thing, I'm not sure we want to be mutating the existing map by calling `.set` - we might want to make a new `Map` and then add the new post to that:

```js
const updatedPostMap = new Map(upvotedPosts)
updatedPostMap.set(upvotedPost.id, upvotedPost)
this.setState({ upvotedPosts: updatedPostMap })
```

something like that, just to avoid mutating an object we're holding in `this.state`.

what do you think?

Reply:
I don't think it would run into the same issue on the homepage and channel page. If you look at the implementation of the `toggleUpvote` function it, after dispatching the async call to the API, dispatches a `SET_POST_DATA` action which will put the updated post record (returned from the API call) into the post reducer. Then the UI on those pages just gets the data out of the post reducer. I don't think that setup is vulnerable to the same race condition, because the data update is done asynchronously after the API action completes for every upvote or downvote that we kick off. 

Here we're not using the post reducer, we're using a different piece of data (on the `postUpvotes` reducer). One way that we could fix things I think would be to look at the `post` reducer for our data instead of the `postUpdate` one, since the `toggleUpvote` function as we're using it here is already going to stick the updated post into the `post` reducer without us doing anything (if you look at the dispatched actions when voting on the search page you can see the `SET_POST_DATA` actions firing).

does that make sense?

Reply:
I think so, thanks!